### PR TITLE
ROX-9824 Triggering evaluation from networkpolicy updates

### PR DESCRIPTION
**Pull request by lvalerom**
## Description

This PR triggers deployment updates when a network policy is updated. For each network policy update we gather all the deployments affected by it, and for each deployment we trigger `ProcessDeployment` in the `Detector`.

**Notice**: the code is protected with the `ROX_NETPOL_FIELDS` feature flag.

## Checklist
- [x] Investigated and inspected CI test results
- [x] Unit test and regression tests added
- ~[ ] Evaluated and added CHANGELOG entry if required~
- ~[ ] Determined and documented upgrade steps~
- ~[ ] Documented user facing changes (create PR based on [stackrox/openshift-docs](https://github.com/stackrox/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))~

If any of these don't apply, please comment below.

## Testing Performed

* Added new unit tests
* Manual testing with local cluster

### CODE REVIEW

File: sensor/kubernetes/listener/resources/convert.go
Comment (vikin91) [inline sensor/kubernetes/listener/resources/convert.go:63]: What would be the advantages of having the `networkPoliciesApplied` stored here?

File: sensor/kubernetes/listener/resources/networkpolicy_test.go
Comment (fredrb) [inline sensor/kubernetes/listener/resources/networkpolicy_test.go:16]: With the way this is now, it seems like the tests aren't running at all in the CI. Maybe add a `SetupTest()` in this setting the feature flag to true so at least we get tests running in `go-unit-tests` step. You can keep this check here to avoid running the tests in release mode.

**Review by vikin91 — CHANGES_REQUESTED** [review]
Looks really good already (very nice set of tests!), there are just several rough edges.

File: sensor/kubernetes/listener/resources/networkpolicy_test.go
Comment (fredrb) [inline sensor/kubernetes/listener/resources/networkpolicy_test.go:27]: nit: I think the tests are a bit easier to read if you define test cases as `map[string]struct{ ... }`.

Then you can write them as:
```go
"New NetworkPolicy": {
	netpol: &storage.NetworkPolicy{
		Id:        "1",
		Namespace: "default",
		Spec: &storage.NetworkPolicySpec{
			PodSelector: &storage.LabelSelector{
				MatchLabels: map[string]string{
					"app":  "sensor",
					"role": "backend",
				},
			},
		},
	},
	oldNetpol: nil,
	action:    central.ResourceAction_CREATE_RESOURCE,
	expectedSelector: []map[string]string{
		{
			"app":  "sensor",
			"role": "backend",
		},
	},
	expectedEmpty: false,
},
```

File: sensor/kubernetes/listener/resources/networkpolicy_test.go
Comment (fredrb) [inline sensor/kubernetes/listener/resources/networkpolicy_test.go:37]: It seems like `Id` and `Namespace` are the same for all the tests cases.

1) Does it make sense to test different values?
2) If not, a nitpick: maybe you can create a function to wrap the creation of test Network Policies? This way all the constant values that are present in all policies won't need to show up in every test case. It makes it easier to read also since the test case will only hold necessary information for this assertion.

File: sensor/kubernetes/listener/resources/networkpolicy_test.go
Comment (fredrb) [inline sensor/kubernetes/listener/resources/networkpolicy_test.go:253]: All these things can be grouped up in a `SetupTest()` if you use testify `suite.Suite`.

File: sensor/kubernetes/listener/resources/networkpolicy_test.go
Comment (fredrb) [inline sensor/kubernetes/listener/resources/networkpolicy_test.go:425]: AFAIK the expectation needs to be defined with a matcher other than the actual value. 
```suggestion
		processDeploymentMock := det.EXPECT().ProcessDeployment(gomock.Any(), gomock.Eq(central.ResourceAction_UPDATE_RESOURCE)).DoAndReturn(func(d *storage.Deployment, _ central.ResourceAction) {
```

File: sensor/kubernetes/listener/resources/networkpolicy_test.go
Comment (fredrb) [inline sensor/kubernetes/listener/resources/networkpolicy_test.go:423]: I'd suggest you wrap all the content inside the `for` loop in a subtest. Otherwise it's difficult to figure out which test failed if there is a failure.
```go
for _, c := range cases {
		t.Run(c.name, func(t *testing.T) {
			// ... inner loop code here
		})
}
```

Plus, that's a very useful usage of the `name` property you already added to the test case struct.

Comment (fredrb) [reply sensor/kubernetes/listener/resources/networkpolicy_test.go:423]: This is valid for the other test as well.

**Review by fredrb — COMMENTED** [review]
Partial review. Mostly looked at the tests. Before continuing with it, I have some questions about the testing strategy:

1) Could we test `ProcessEvent` instead of the internal methods of the dispatcher (`getSelector` and `updateDeploymentsFromStore`)? I think it makes for better tests because the test execution would look similar to how other caller would interact with it.
2) I noticed that you are using the original store implementations for both deployments and network policies. Wouldn't the tests be easier to implement if you used their mocks? I think it ultimately depends on how "isolated" this part of the code is from the stores. If you think that the store implementation is relevant to be tested here as well then I'm fine with how it is. I wrote the tests for #1176 using a mocked `store.NetworkPolicyStore` and it was pretty simple to set it up. 
```go
suite.networkStore.EXPECT().
		Find(gomock.Eq(ns), gomock.Eq(labels)).
		Return(testCase.policiesInStore)
```

File: sensor/kubernetes/listener/resources/networkpolicy_test.go
Comment (fredrb) [reply sensor/kubernetes/listener/resources/networkpolicy_test.go:17]: +1

File: sensor/kubernetes/listener/resources/networkpolicy.go
Comment (lvalerom) [reply sensor/kubernetes/listener/resources/networkpolicy.go]: I've changed it so it uses the second parameter of `ProcessEvent`.

File: sensor/kubernetes/listener/resources/networkpolicy.go
Comment (lvalerom) [reply sensor/kubernetes/listener/resources/networkpolicy.go:106]: 1. At first I thought of using [Empty()](https://github.com/kubernetes/apimachinery/blob/6550efdb74445c227aa0bd74017b90a9a8cd07e2/pkg/labels/selector.go#L54). The problem; it returns `false` when the selector is `nothingSelector` which we use [here](https://github.com/stackrox/stackrox/blob/aad0c634907b667353acb6a31db1a3b34ad514b9/sensor/kubernetes/listener/resources/selector.go#L33). Also using the `len` of the selector does not work because if we are in the case of an old network policy with no selector and we update the `PodSelector`, we will have `selectorDisjunction` like `[{}, {"key1: "value1", "key2": "value2"}] and this should be considered as empty.
2. Right now, if the map is empty we create a `nothingSelector` which returns always `false` so [here](https://github.com/stackrox/stackrox/blob/9575e3acc469acfac27b290f5d92c4bb7b473e0f/sensor/kubernetes/listener/resources/deployment_store.go#L83-L85) we will never match anything when we should match everything.

File: sensor/kubernetes/listener/resources/networkpolicy_test.go
Comment (lvalerom) [reply sensor/kubernetes/listener/resources/networkpolicy_test.go]: I believe so, I've changed the `return` for `t.Skip`

File: sensor/kubernetes/listener/resources/convert.go
Comment (lvalerom) [reply sensor/kubernetes/listener/resources/convert.go:63]: If we decide to have it there, later in the `enricher` we only need to access this field. It's a matter of preference on where to put the code right now.

In the future, storing the `networkPoliciesApplied` (somewhere) will allow us to compare changes between how the deployment was affected and how it is affected now.

File: sensor/kubernetes/listener/resources/convert.go
Comment (lvalerom) [reply sensor/kubernetes/listener/resources/convert.go:62]: Done. I'll mark the `TODO` in the next commit.

File: sensor/kubernetes/listener/resources/networkpolicy_test.go
Comment (lvalerom) [reply sensor/kubernetes/listener/resources/networkpolicy_test.go]: Done. I've changed the tests to use testify and now everything should be set in `SetupTest()`

File: sensor/kubernetes/listener/resources/networkpolicy_test.go
Comment (lvalerom) [reply sensor/kubernetes/listener/resources/networkpolicy_test.go]: Done. Thanks for the tip, it looks more readable this way.

File: sensor/kubernetes/listener/resources/networkpolicy_test.go
Comment (lvalerom) [reply sensor/kubernetes/listener/resources/networkpolicy_test.go]: In some tests the `namespace` is different but I still think that your suggestion of wrapping the creation of `storage.NetworkPolicy`s is valid.

File: sensor/kubernetes/listener/resources/networkpolicy_test.go
Comment (lvalerom) [reply sensor/kubernetes/listener/resources/networkpolicy_test.go]: Done

File: sensor/kubernetes/listener/resources/networkpolicy_test.go
Comment (lvalerom) [reply sensor/kubernetes/listener/resources/networkpolicy_test.go]: Oops... I've fixed it 😅

File: sensor/kubernetes/listener/resources/networkpolicy_test.go
Comment (lvalerom) [reply sensor/kubernetes/listener/resources/networkpolicy_test.go]: Yeah, you're right. I've implemented this in both tests.

**Comment by lvalerom** [comment]
> Partial review. Mostly looked at the tests. Before continuing with it, I have some questions about the testing strategy:
> 
> 1. Could we test `ProcessEvent` instead of the internal methods of the dispatcher (`getSelector` and `updateDeploymentsFromStore`)? I think it makes for better tests because the test execution would look similar to how other caller would interact with it.
> 2. I noticed that you are using the original store implementations for both deployments and network policies. Wouldn't the tests be easier to implement if you used their mocks? I think it ultimately depends on how "isolated" this part of the code is from the stores. If you think that the store implementation is relevant to be tested here as well then I'm fine with how it is. I wrote the tests for #1176 using a mocked `store.NetworkPolicyStore` and it was pretty simple to set it up.
> 
> ```go
> suite.networkStore.EXPECT().
> 		Find(gomock.Eq(ns), gomock.Eq(labels)).
> 		Return(testCase.policiesInStore)
> ```

* `ProcessEvent` always returns the same and all the logic that we want to test is internal. I think we should keep the tests separate and add tests for `ProcessEvent` to assert the insertions in the `NetworkPolicyStore`.
* Following on that, right now pretty much all the code in the `resources` package uses the `DeploymentStore` not the interface in `store.DeploymentStore` so we cannot use the mock here. I think is a good idea to change this put I would do it in a separate PR because it's going to change a lot of files in the `resources` package.

I would also add the tests to `ProcessEvent` in a separate PR once #1176 is merged since the mock for the `NetworkPolicyStore` is already implemented there.

What do you think? If you are ok with it I can open following tasks for these two things.

File: sensor/kubernetes/listener/resources/networkpolicy.go
Comment (vikin91) [reply sensor/kubernetes/listener/resources/networkpolicy.go:106]: I see. So involving `nothingSelector` makes it clear that empty selector matches everything and nothing selector matches nothing. I am thinking how to make it clearer when reading the code 🤔  Maybe we could rename `isEmptySelector` into `matchesAll` or `matchAll`?

Or maybe split it into two separate functions - one for all deployments `functionName(np *storage.NetworkPolicy)` and the other `function2(np *storage.NetworkPolicy, sel selector)`?

**Comment by fredrb** [comment]
> What do you think? If you are ok with it I can open following tasks for these two things.

That works for me. Thanks @lvalerom

File: sensor/kubernetes/listener/resources/networkpolicy_test.go
Comment (vikin91) [reply sensor/kubernetes/listener/resources/networkpolicy_test.go]: This is cool to use map instead of array! We actually loose the order of execution (it is random for map), but it doesn't play a role here.

File: sensor/kubernetes/listener/resources/networkpolicy.go
Comment (lvalerom) [reply sensor/kubernetes/listener/resources/networkpolicy.go:106]: I like the renaming it to `matchesAll` approach. I wouldn't separate the logic in two functions because in the future we might need to add some logic in loops (i.e if we store the augmented object in the deployment wrap) and I think is better if we don't have duplicated code. At that point I would probably join both loops into one.
```go
var deployments []*deploymentWrap
if matchesAll {
    deployments = h.deploymentStore.getAllDeploymentsInNamespace(np.GetNamespace())
} else {
    deployments = h.deploymentStore.getMatchingDeployments(np.GetNamespace(), sel)
}
for _, d := range deployments {
...
}
```

What do you think?

Comment (vikin91) [reply sensor/kubernetes/listener/resources/networkpolicy.go:106]: Yes, sounds good - this should be enough to get what is happening here. You may optionally add a comment explaining why we use it (and mention `nothingSelector`) - up to you.

Comment (lvalerom) [reply sensor/kubernetes/listener/resources/networkpolicy.go]: Done

**Review by vikin91 — APPROVED** [review]
I think it looks good already - thanks for improving!

File: sensor/kubernetes/listener/resources/networkpolicy.go
Comment (fredrb) [inline sensor/kubernetes/listener/resources/networkpolicy.go:28]:
```suggestion
// ProcessEvent processes a network policy resource event and returns the sensor events to generate.
```

File: sensor/kubernetes/listener/resources/networkpolicy.go
Comment (fredrb) [inline sensor/kubernetes/listener/resources/networkpolicy.go:61]: I think there is some potential for improvement in this function with regards to returning both a selector and a `matchAll` boolean value. A few things that I thought:

1. I understand that `nothingSelector` will not match anything if used as a matcher, but couldn't we simply check introduce a new method (as opposed to `SelectorFromMap`) that will return `labels.Everything()` instead?
2. If old policy is a match-all policy we can early return, since whatever the new policy matches is irrelevant, we'll anyway select all deployments in the namespace.
3. Consider that if no old policy is provided we can start the selector from a `labels.Nothing()` and combine that with `or(...)`. This would simplify the second portion of the code. 

I tried exercising this below but didn't test it since the unit tests are expecting the `matchAll` returns. This comes back to my previous comment about testing internal functions. I understand why you did it like this, but the flip side is that it's difficult to refactor code without having to refactor the tests. 😞 

```go
func matcherOrEverything(labelsMap map[string]string) labels.Selector {
	if len(labelsMap) == 0 {
		return labels.Everything()
	}
	return labels.SelectorFromSet(labels.Set(labelsMap))
}

func (h *networkPolicyDispatcher) getSelector(np, oldNp *storage.NetworkPolicy, action central.ResourceAction) selector {
	var sel selector
	if oldNp != nil {
		sel = matcherOrEverything(oldNp.GetSpec().GetPodSelector().GetMatchLabels())	
                // potential early return but not necessary
		if sel == labels.Everything() {
			return sel
		}	
	} else {
		sel = labels.Nothing()
	}

	matchLabels := np.GetSpec().GetPodSelector().GetMatchLabels()
	if action == central.ResourceAction_UPDATE_RESOURCE {
		sel = or(sel, matcherOrEverything(matchLabels))
	} else if action == central.ResourceAction_CREATE_RESOURCE {
		sel = matcherOrEverything(matchLabels)
	}	
	
	return sel
}
```

**Review by fredrb — COMMENTED** [review]
I'm unblocking this, but I think we should do the refactor I mentioned. I'll leave it up to you if you want to do it in a follow-up PR or here.

File: sensor/kubernetes/listener/resources/networkpolicy.go
Comment (fredrb) [reply sensor/kubernetes/listener/resources/networkpolicy.go:61]: It could even be simpler if we ignore the `action` check, which won't make a difference if we're combining with `or labels.Nothing()`:

```go
func (h *networkPolicyDispatcher) getSelector(np, oldNp *storage.NetworkPolicy, action central.ResourceAction) selector {
	var sel selector
	if oldNp != nil {
		sel = matcherOrEverything(oldNp.GetSpec().GetPodSelector().GetMatchLabels())	
                // potential early return but not necessary
		if sel == labels.Everything() {
			return sel
		}	
	} else {
		sel = labels.Nothing()
	}

	matchLabels := np.GetSpec().GetPodSelector().GetMatchLabels()
        return or(sel, matcherOrEverything(matchLabels))
}
```

File: sensor/kubernetes/listener/resources/networkpolicy.go
Comment (vikin91) [inline sensor/kubernetes/listener/resources/networkpolicy.go:74]: I have just noticed that we have hardcoded action `central.ResourceAction_UPDATE_RESOURCE` here (no matter whether we upsert or delete). `ProcessDeployment` behaves differently for `REMOVE_RESOURCE` - are we sure that this value can be hardcoded?

**Review by vikin91 — COMMENTED** [review]
It looks very good! 
I just found one potential issue that we maybe should check before we merge - see comment.

File: sensor/kubernetes/listener/resources/networkpolicy.go
Comment (fredrb) [reply sensor/kubernetes/listener/resources/networkpolicy.go:74]: I think it has to be `update`, no? Because we're reacting to adding, updating, or removing a `NetworkPolicy`. So regardless of what action is, if it was associated with that deployment, we need to pretend that the deployment was updated in order to trigger the evaluation.

Comment (vikin91) [reply sensor/kubernetes/listener/resources/networkpolicy.go:74]: I looked at `processDeploymentNoLock` and I think we are safe. It would be important if we would remove a deployment - then we would need to close violations for it. But as we only change the NetworkPolicies, then the only action making sense is UPDATE.

**Review by vikin91 — APPROVED** [review]

**Review by fredrb — APPROVED** [review]
Didn't find any blockers. Looks good to go! Thanks for making the refactor in this PR, I think the code looks much cleaner now.

**Pull request merged**
